Add special faster handling of IN when checking constant collections

systay opened the pull request:
test-query:

```
MATCH (a)-[r]->(b) 
WHERE id(a) IN [$idRange] AND id(b) IN [$idRange] 
RETURN r
```

`$idRange` is a string with 100 numbers in it

Small database with 10k nodes, and 1,5M relationships, on my computer, I get these numbers. This is in ms taken per query. First result is always slower because it loads and compiles a bunch of code and works off cold caches.

without fast IN checking:

```
2722
540
533
517
540
528
513
507
518
515
522
```

With fast-in

```
2194
57
52
38
29
28
34
30
31
27
28
```

Commit:
Add special faster handling of IN when checking constant collections

Commit:
Make sure ConstantIn handles null and lists-in-lists well

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/predicates/ConstantIn.scala:
```diff
@@ -0,0 +1,56 @@
+
+It uses a cache for the <constant-expression> value, and turns it into a Set, for fast existence checking
+ */
+case class ConstantIn(value: Expression, list: Expression) extends Predicate with CollectionSupport {
```
Should call it @lutovich instead :)

Commit:
Minor test code cleanup

Removed a test expecting indexes to be used
even when the predicate will never evaluate to true

Changed a test fixture to run assertions as individual
tests to make it easier to see what is failing and what works

Removed a test added by mistake

Commit:
indentation

Commit:
Solve test failures

* Make sure indexes can be used with IN expressions
* Rewrite IN against empty collections to literal false
* Remove invalid test
